### PR TITLE
Initial Take on UX Guidelines: Principles and Patterns

### PR DESCRIPTION
**Pull request by stevegrossi**
My initial reason for adding this section was because we have a lot of ways of doing things in the app, some of which we’re trying to phase out as soon as possible, and some of which we’re trying to phase in. The second “Preferred Patterns” section encompasses that goal. I hope we’ll continue to add to it as we learn.

But I also realized that some general principles can help guide our decisions as well, especially as we scale our team.

I drew the most inspiration from this Quora question: https://www.quora.com/What-should-a-UX-style-guide-contain

### CODE REVIEW

**Comment by stevegrossi** [comment]
@lessonly/developers: thoughts and additions?

File: ux-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+## Principles
+
+- **Be consistent.** This is the most important rule of interface design. Things that are similar conceptually should look and work similarly. Things which are different conceptually should look and work differently. For example, adding tags to lessons should work just like adding tags to courses. Likewise, an error message should not be the same color as a success message.
+- **Always provide feedback.** Take the time to handle error cases and provide human-readable feedback to the user. An app not working because a system is down is annoying; an app not working for no apparent reason at all is maddening.
+- **Prevent errors in addition to handling them.** Certainly provide a helpful error message if a form is submitted with invalid input, but consider preventing invalid input in the first place. For example, if a phone number in a certain format is required, use the `required`, `placeholder="123-456-7890"`, and `pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"` attributes.
```
Comment (armilam) [inline ux-guidelines.md]: This flipside of this is not over-validating input. If someone has an international phone number, we should allow those as well.

File: ux-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+## Preferred Patterns
+
+These are patterns we may not use everywhere yet, but which we want to adopt.
+
+- **Edit-in-Place**: With this pattern, initiating an edit replaces the content itself with a form. Upon submission, the updated content once again replaces the form. This avoids context-switching (not to mention page reloads) which slows users down. Keep in mind:
+  - The editable form should be similar in size and style to the content itself, in order to reinforce that they correspond to the same thing.
+  - Pre-select the text in the form input so users can immediately begin typing.
+  - Keyboard accessability is especially helpful here. `Escape` should cancel, while `Enter`/`Return` should submit the update (except when dealing with multiline text).
```
Comment (armilam) [inline ux-guidelines.md]: *accessibility*

**Comment by armilam** [comment]
This is great! I look forward to adding to this.

File: ux-guidelines.md (hunk 1 above)
Comment (stevegrossi) [inline ux-guidelines.md]: Agreed. Perhaps we should add some version of [the robustness principle](https://en.wikipedia.org/wiki/Robustness_principle)?

> Be conservative in what you do, be liberal in what you accept from others.

File: ux-guidelines.md (hunk 2 above)
Comment (stevegrossi) [inline ux-guidelines.md]: Ah geez, after that whole other section about using correct English... :wink:

**Comment by stevegrossi** [comment]
I've decided to make this a lesson: https://about.lesson.ly/preview/lesson/23095-user-experience-guidelines. Let's continue any further discussion in Flowdock.

**Pull request closed without merging**